core/src/impl/Kokkos_Default_Graph_Impl.hpp

```diff
@@ -141,16 +141,29 @@ struct GraphImpl : private ExecutionSpaceInstanceStorage<ExecutionSpace> {
     m_has_been_instantiated = true;
   }
 
-  // FIXME The execution space instance should be passed to the nodes.
-  void submit(const ExecutionSpace& /* exec */) {
+  void submit(const ExecutionSpace& exec) {
     if (!m_has_been_instantiated) instantiate();
     // This reset is gross, but for the purposes of our simple host
     // implementation...
     for (auto& sink : m_sinks) {
       sink->reset_has_executed();
     }
+
+    // We don't know where the nodes will execute, so we need to fence the given
+    // execution space instance before proceeding. This is the simplest way
+    // of guaranteeing that the kernels in the graph are correctly "enqueued".
+    exec.fence(
+        "Kokkos::DefaultGraph::submit: fencing before launching graph nodes");
+
     for (auto& sink : m_sinks) {
-      sink->execute_node();
+      sink->execute_node(exec);
+    }
+
+    // Once all sinks have been executed, we need to fence them.
+    for (const auto& sink : m_sinks) {
+      if (sink->awaitable())
+        sink->get_execution_space().fence(
+            "Kokkos::DefaultGraph::submit: fencing before ending graph submit");
```

Comment:
We only kinda need to do that if they are different from the one passed to the graph submit or?

Reply:
You might possibly be right. I need to double-check.

Reply:
Indeed, you're right.

I thought about it with 2 though-experiments.

# Case 1 - only sinks

There are only sinks (*i.e.* they don't have predecessors). Then, it is true that we need to fence only the sinks that where not executing on the passed `exec`.

# Case 2 - complex DAG

The sinks should already have fenced their predecessors, so we come back to the situation of case 1.


Conclusion: you must be right. I'll do it.

Reply:
@crtrott Done, feel free to close the thread.